### DISCUSSION TITLE
Phase Power not reporting

### DISCUSSION THREAD

**Discussion by burnski666** [emporia-vue-local/esphome#207] [Q&A]
verified all of the clamps are identified as on the correct phases, all of my 16 sensors are reporting accurate power. The main clamps are reporting voltage & frequency, ive done the calibration and its reporting within 0.1v. However Im not getting the Phase_a, Phase_b, balance & total power readings. Heres the log: [logs_emporia-vue_logs.txt](https://github.com/emporia-vue-local/esphome/files/12641822/logs_emporia-vue_logs.txt)

 

![Screenshot 2023-09-16 205913](https://github.com/emporia-vue-local/esphome/assets/13058460/498e0dca-bc7f-43fb-9f8b-05ed3db6e6ec)

and my YAML:

```

esphome:
  name: emporiavue2
  friendly_name: vue2

external_components:
  - source: github://emporia-vue-local/esphome@dev
    components:
      - emporia_vue

esp32:
  board: esp32dev
  framework:
    type: esp-idf
    version: recommended

# Enable Home Assistant API
api:
  encryption:
    # Encryption key is generated by the new device wizard.
 
  services:
    - service: play_rtttl
      variables:
        song_str: string
      then:
        - rtttl.play:
            rtttl: !lambda 'return song_str;'

ota:
  # Create a secure password for pushing OTA updates.
 
# Enable logging
logger:
  logs:
    # by default, every reading will be printed to the UART, which is very slow
    # This will disable printing the readings but keep other helpful messages
    sensor: INFO

wifi:
  # Wifi credentials are stored securely by new device wizard.
  ssid: !secret wifi_ssid
  password: !secret wifi_password

preferences:
  # the default of 1min is far too short--flash chip is rated
  # for approx 100k writes.
  flash_write_interval: "48h"

output:
  - platform: ledc
    pin: GPIO12
    id: buzzer
  - platform: gpio
    pin: GPIO27
    id: buzzer_gnd

rtttl:
  output: buzzer
  on_finished_playback:
    - logger.log: 'Song ended!'

button:
  - platform: template
    name: "Two Beeps"
    on_press:
      - rtttl.play: "two short:d=4,o=5,b=100:16e6,16e6"

light:
  - platform: status_led
    name: "D3_LED"
    pin: 23
    restore_mode: ALWAYS_ON
    entity_category: config

i2c:
  sda: 21
  scl: 22
  scan: false
  frequency: 200kHz  # recommended range is 50-200kHz
  id: i2c_a

time:
  - platform: sntp
    id: my_time

# these are called references in YAML. They allow you to reuse
# this configuration in each sensor, while only defining it once
.defaultfilters:
  - &throttle_avg
    # average all raw readings together over a 5 second span before publishing
    throttle_average: 5s
  - &throttle_time
    # only send the most recent measurement every 60 seconds
    throttle: 60s
  - &invert
    # invert and filter out any values below 0.
    lambda: 'return max(-x, 0.0f);'
  - &pos
    # filter out any values below 0.
    lambda: 'return max(x, 0.0f);'
  - &abs
    # take the absolute value of the value
    lambda: 'return abs(x);'

sensor:
  - platform: emporia_vue
    i2c_id: i2c_a
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK  # Vue device wire color
        calibration: 0.0229847  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "Phase A Voltage"
          filters: [*throttle_avg, *pos]
        frequency:
          name: "Phase A Frequency"
          filters: [*throttle_avg, *pos]
      - id: phase_b  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: RED  # Vue device wire color
        calibration: 0.0217332  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "Phase B Voltage"
          filters: [*throttle_avg, *pos]
        phase_angle:
          name: "Phase B Phase Angle"
          filters: [*throttle_avg, *pos]
    ct_clamps:
      - phase_id: phase_a
        input: "A"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_a_power
          filters: [*pos]
      - phase_id: phase_b
        input: "B"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_b_power
          filters: [*pos]
      # Pay close attention to set the phase_id for each breaker by matching it to the phase/leg it connects to in the panel
      - { phase_id: phase_a, input:  "1", power: { id:  cir1, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "2", power: { id:  cir2, filters: [ *pos, multiply: 2 ] } }
      - { phase_id: phase_b, input:  "3", power: { id:  cir3, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "4", power: { id:  cir4, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "5", power: { id:  cir5, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "6", power: { id:  cir6, filters: [ *pos ] } }
      - { phase_id: phase_b, input:  "7", power: { id:  cir7, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "8", power: { id:  cir8, filters: [ *pos ] } }
      - { phase_id: phase_a, input:  "9", power: { id:  cir9, filters: [ *pos ] } }
      - { phase_id: phase_b, input: "10", power: { id: cir10, filters: [ *pos ] } }
      - { phase_id: phase_a, input: "11", power: { id: cir11, filters: [ *pos ] } }
      - { phase_id: phase_a, input: "12", power: { id: cir12, filters: [ *pos ] } }
      - { phase_id: phase_b, input: "13", power: { id: cir13, filters: [ *pos ] } }
      - { phase_id: phase_a, input: "14", power: { id: cir14, filters: [ *pos ] } }
      - { phase_id: phase_b, input: "15", power: { id: cir15, filters: [ *pos ] } }
      - { phase_id: phase_a, input: "16", power: { id: cir16, filters: [ *pos, multiply: 2 ] } }
    on_update:
      then:
        - component.update: total_power
        - component.update: balance_power
  - { platform: copy, name: "Phase A Power", source_id: phase_a_power, filters: *throttle_avg }
  - { platform: copy, name: "Phase B Power", source_id: phase_b_power, filters: *throttle_avg }
  - { platform: copy, name: "Total Power", source_id: total_power, filters: *throttle_avg }
  - { platform: copy, name: "Balance Power", source_id: balance_power, filters: *throttle_avg }
  - { platform: copy, name: "Furnace", source_id:  cir1, filters: *throttle_avg }
  - { platform: copy, name: "A/C", source_id:  cir2, filters: *throttle_avg }
  - { platform: copy, name: "Water Heater, Mini Fridge", source_id:  cir3, filters: *throttle_avg }
  - { platform: copy, name: "Microwave", source_id:  cir4, filters: *throttle_avg }
  - { platform: copy, name: "Pool Shed", source_id:  cir5, filters: *throttle_avg }
  - { platform: copy, name: "Sump Pump", source_id:  cir6, filters: *throttle_avg }
  - { platform: copy, name: "Deep Freezer", source_id:  cir7, filters: *throttle_avg }
  - { platform: copy, name: "Washing Machine", source_id:  cir8, filters: *throttle_avg }
  - { platform: copy, name: "Refridgerator", source_id:  cir9, filters: *throttle_avg }
  - { platform: copy, name: "Kitchen", source_id: cir10, filters: *throttle_avg }
  - { platform: copy, name: "Dishwasher", source_id: cir11, filters: *throttle_avg }
  - { platform: copy, name: "Den, Sitting Room TV's", source_id: cir12, filters: *throttle_avg }
  - { platform: copy, name: "Clothes Dryer", source_id: cir13, filters: *throttle_avg }
  - { platform: copy, name: "Garage", source_id: cir14, filters: *throttle_avg }
  - { platform: copy, name: "Network", source_id: cir15, filters: *throttle_avg }
  - { platform: copy, name: "Garage Heater", source_id: cir16, filters: *throttle_avg }
  - platform: template
    lambda: return id(phase_a_power).state + id(phase_b_power).state;
    update_interval: never   # will be updated after all power sensors update via on_update trigger
    id: total_power
    device_class: power
    state_class: measurement
    unit_of_measurement: "W"
  - platform: total_daily_energy
    name: "Total Daily Energy"
    power_id: total_power
    accuracy_decimals: 0
    filters: *throttle_time
  - platform: template
    lambda: !lambda |-
      return max(0.0f, id(total_power).state -
        id( cir1).state -
        id( cir2).state -
        id( cir3).state -
        id( cir4).state -
        id( cir5).state -
        id( cir6).state -
        id( cir7).state -
        id( cir8).state -
        id( cir9).state -
        id(cir10).state -
        id(cir11).state -
        id(cir12).state -
        id(cir13).state -
        id(cir14).state -
        id(cir15).state -
        id(cir16).state);
    update_interval: never   # will be updated after all power sensors update via on_update trigger
    id: balance_power
    device_class: power
    state_class: measurement
    unit_of_measurement: "W"
  - platform: total_daily_energy
    name: "Balance Daily Energy"
    power_id: balance_power
    accuracy_decimals: 0
    filters: *throttle_time
  - { power_id:  cir1, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 1 Daily Energy", filters: *throttle_time }
  - { power_id:  cir2, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 2 Daily Energy", filters: *throttle_time }
  - { power_id:  cir3, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 3 Daily Energy", filters: *throttle_time }
  - { power_id:  cir4, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 4 Daily Energy", filters: *throttle_time }
  - { power_id:  cir5, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 5 Daily Energy", filters: *throttle_time }
  - { power_id:  cir6, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 6 Daily Energy", filters: *throttle_time }
  - { power_id:  cir7, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 7 Daily Energy", filters: *throttle_time }
  - { power_id:  cir8, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 8 Daily Energy", filters: *throttle_time }
  - { power_id:  cir9, platform: total_daily_energy, accuracy_decimals: 0, name:  "Circuit 9 Daily Energy", filters: *throttle_time }
  - { power_id: cir10, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 10 Daily Energy", filters: *throttle_time }
  - { power_id: cir11, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 11 Daily Energy", filters: *throttle_time }
  - { power_id: cir12, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 12 Daily Energy", filters: *throttle_time }
  - { power_id: cir13, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 13 Daily Energy", filters: *throttle_time }
  - { power_id: cir14, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 14 Daily Energy", filters: *throttle_time }
  - { power_id: cir15, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 15 Daily Energy", filters: *throttle_time }
  - { power_id: cir16, platform: total_daily_energy, accuracy_decimals: 0, name: "Circuit 16 Daily Energy", filters: *throttle_time }
    

```

**Comment by flaviut** [comment]
Are your main clamps backwards? What do you see when you remove the `*pos` on the main clamps, i.e.

```
    ct_clamps:
      - phase_id: phase_a
        input: "A"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_a_power
          filters: []
      - phase_id: phase_b
        input: "B"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_b_power
          filters: []
```

**Comment by burnski666** [comment]
> Are your main clamps backwards? What do you see when you remove the `*pos` on the main clamps, i.e.
> 
> ```
>     ct_clamps:
>       - phase_id: phase_a
>         input: "A"  # Verify the CT going to this device input also matches the phase/leg
>         power:
>           id: phase_a_power
>           filters: []
>       - phase_id: phase_b
>         input: "B"  # Verify the CT going to this device input also matches the phase/leg
>         power:
>           id: phase_b_power
>           filters: []
> ```




They are reporting but negative now. I assume that means I have my main CT's backwards?


![Screenshot 2023-09-17 143109](https://github.com/emporia-vue-local/esphome/assets/13058460/54f57071-741e-405e-9553-fcef19d41bde)

Any way to tweak my yaml to work backwards? Kinda nervous around those main lines that cant be turned off...

**Reply by flaviut** [reply]
yup, `*invert`

**Reply by burnski666** [reply]
amazing!

![Screenshot 2023-09-17 144319b](https://github.com/emporia-vue-local/esphome/assets/13058460/abce7ef5-4174-4756-a11e-10f167dc28fe)

**Comment by eprochasson** [comment]
Hi,

I had a similar issue, although I'm fairly certain my clamps are put the right way (and didn't have any issue with the Emporia app). The fix works great though!

**Reply by mwaterbu** [reply]
I'm also having this same issue, but my clamps are definitely not on backwards.

~~Could this be an issue with this project?~~

**Reply by eprochasson** [reply]
So I dug deeper into this, and other threads talk about variation in the Vue2 model (some info about black/red cables and what not). See for example https://github.com/emporia-vue-local/esphome/discussions/82.

I must confess that I don't understand the problems entirely, however I was able to fix it by:
- adding the *invert filter for the mains
- double checking which main is on which phase (by double checking, I mean tweaking the configuration file until it gave reasonable values)
- double checking all the input phase as well

I now have credible numbers and a balanced balance (in my case, I expect some discrepancy, as I have sub panels for which I can only monitor one load. I use a *2 multiplier but that's not a very good approximation since the load on each phase would depend on the actual load in the subpanels).

**Reply by mwaterbu** [reply]
I initially "fixed" mine with the ```*invert``` filter as well, but then I realized that all of my loads were showing on the wrong phase (ie. clamp 1, which is measuring a circuit on phase A would show the expected value, but the power would show on phase B instead of A).
I tried swapping the input colors for my phases (and removed ```*invert```), which fixed the phase powers, but then all of the individual circuits showed 0W, so I would have needed to invert all those or swap them to the incorrect phase to get them right.
The solution for me was to swap "A" and "B" for the ```ct_clamps``` phase inputs. I suppose physically swapping the 3.5mm inputs would also work.

**Reply by mwaterbu** [reply]
> I now have credible numbers and a balanced balance (in my case, I expect some discrepancy, as I have sub panels for which I can only monitor one load. I use a *2 multiplier but that's not a very good approximation since the load on each phase would depend on the actual load in the subpanels).

Assuming you have split phase (120/240V), you have a couple options:

1. Clamp one leg, and use a multiplier of 2, as you have done. However, as you have also noted, this is inaccurate for unbalanced loads.
2. If the load is not balanced, use two clamps, and measure each leg individually.
3. If the load is not balanced, you can also pass both wires through a single clamp, but they need to be passing through in opposite directions, or you will just get the delta between them. There is a more thorough description and picture of that [here](https://docs.iotawatt.com/en/master/CTbasics.html#v-circuits).

Basically, if the load is balanced, just clamp one, and use a multiplier of 2 (option 1). If unbalanced, and you have clamps to spare, or your wires are too thick to pass both through one clamp, use two clamps (option 2). If unbalanced, and you either do not have clamps to spare, or you prefer to see the load as one entity, pass both legs through one clamp (option 3). With option 2 or 3, make sure you remove the multiplier.

Subpanels will not be balanced as you have noted, but there are other cases too. I had to use option 3 on my electric range, as I observed that there are 120V loads for the auxiliary functions (controls, oven light, etc), and the multiplier doubled those, causing my load to be off.

**Comment by gonzalop** [comment]
To add one more data point, I've setup 3 Vue2 in three separate panels. All had the 200A CTs arrow pointing in the right direction but the 2 clamps in one of the panels were reporting negative values, which I fixed by using *invert. It happens to be the first Vue I purchased (circa November 2021)

The emporia app was working as expected even though I had several 50A clamps pointing the wrong way in that same panel.

**Comment by ZinTheDog** [comment]
I believe I discovered the root cause of the issue!  Initially, I had used the *invert filter and that worked, but like mwaterbu I noticed power was reporting on the wrong Phase so I reverted back and swapped the Phase A and Phase B CT clamps.  The reason this worked was because the Phase A and Phase B lines "cross" in the panel breaker and it wasn't obvious from the outside.  I did a continuity test to see which backplane each feed line was truly connected to (I am able to turn the power off entirely to my panel feed lines).

In my original install I had incorrectly assumed the line on the right fed the backplane on the right and that the line on the left fed the backplane on the left.  That was my mistake, the line on the right feeds the left back plane and the line on the left feeds the right back plane in my instance.

**Reply by mwaterbu** [reply]
I went back and checked, and have the exact same situation (the phase lines "cross" in the main breaker), which explains the discrepancy I was seeing. It is a bit strange though as schematic on the panel cover does not indicate this, unless I have misread it.

**Comment by nrobfd** [comment]
I'm having a similar issue. I have 3 Vue2 devices in one panel so that I can monitor every circuit. I only have the main ct clamps plugged in for Device 1. Initially they were reading 0. But I switched inputs A and B in the configuration file.

Now, If I total all the power readings for the CT circuit clamps, I get 1716.4 watts on Phase A and 377.2 watts on Phase B. But the readings from the Main CT Clamps are 1676.7 watts on Phase A and 480.6 watts on Phase B. Is this discrepancy expected / within the range of error on these devices or do I have something wrong in my configuration?


Note: I do have a 60amp unmonitored breaker in this panel feeding a pool sub panel, but it's current power draw is minimal as the pool is winterized and the only thing currently powered is monitored by a smart plug and is drawing little to number power.
